core/unix/signal.c

```diff
@@ -4838,12 +4853,24 @@ master_signal_handler_C(byte *xsp)
                 dcontext = NULL;
             }
         }
-        if (dcontext == NULL) {
-            /* Check for a temporarily-native thread we're synch-ing with. */
-            tr = thread_lookup(get_sys_thread_id());
-            if (tr != NULL)
-                dcontext = tr->dcontext;
-        }
+    }
+    if (dcontext == NULL &&
+        /* Check for a temporarily-native thread we're synch-ing with. */
+        (sig == SUSPEND_SIGNAL
+#ifdef X86
+         || (INTERNAL_OPTION(safe_read_tls_init) &&
+             /* Check for whether this is a thread that makes a clone, in which case its
+              * magic field is temporarily invalid. It is also possible that it is a new
+              * thread on its way to init, but in this case the thread will not yet be
+              * known, because the thread will be added to the all_threads list after its
+              * tls has been initialized (i#2921).
```

Comment:
I don't think this comment captures what we're doing because it doesn't talk about lock safety or native cases beyond clone.  We're trying to handle a signal in a temporarily-native thread, but it is unsafe to call thread_lookup from an arbitrary signal due to the lock it acquires.  If the magic field has our invalid sentinel we assume it's safe b/c it means the thread is either in app code (for native via API like DR_EMIT_GO_NATIVE) or in the clone window (where we know it doesn't grab a problematic lock from inspection of our own code).  Please generalize the comment: this is not just about a clone or init, and it should explain the lock problem.

Reply:
Until a few minutes ago, I didn't know DR_EMIT_GO_NATIVE exists. I assume a temporarily-native thread has its tls magic field still set to valid?

Reply:
Comment refined.